// @metaplex-foundation/js/dist/esm/plugins/nftModule/approveNftCollectionAuthority.mjs
import { createApproveCollectionAuthorityInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { SystemProgram } from '@solana/web3.js';
import { toMintAddress } from './helpers.mjs';
import { findMetadataPda, findCollectionAuthorityRecordPda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _approveNftCollectionAuthorityClient(nftOrSft, collectionAuthority, input = {}) {
  return this.metaplex.operations().getTask(approveNftCollectionAuthorityOperation({ ...input,
    mintAddress: toMintAddress(nftOrSft),
    collectionAuthority
  }));
}
/** @internal */

function _approveNftCollectionAuthorityBuildersClient(input) {
  return approveNftCollectionAuthorityBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'ApproveNftCollectionAuthorityOperation';
const approveNftCollectionAuthorityOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const approveNftCollectionAuthorityOperationHandler = {
  handle: async (operation, metaplex) => {
    return approveNftCollectionAuthorityBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const approveNftCollectionAuthorityBuilder = (metaplex, params) => {
  var _params$systemProgram, _params$instructionKe;

  const {
    mintAddress,
    collectionAuthority,
    updateAuthority = metaplex.identity(),
    payer = metaplex.identity()
  } = params;
  const metadata = findMetadataPda(mintAddress);
  const collectionAuthorityRecord = findCollectionAuthorityRecordPda(mintAddress, collectionAuthority);
  return TransactionBuilder.make().setFeePayer(payer) // Approve the collection authority.
  .add({
    instruction: createApproveCollectionAuthorityInstruction({
      collectionAuthorityRecord,
      newCollectionAuthority: collectionAuthority,
      updateAuthority: updateAuthority.publicKey,
      payer: payer.publicKey,
      metadata,
      mint: mintAddress,
      systemProgram: (_params$systemProgram = params.systemProgram) !== null && _params$systemProgram !== void 0 ? _params$systemProgram : SystemProgram.programId
    }),
    signers: [payer, updateAuthority],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'approveCollectionAuthority'
  });
};

export { _approveNftCollectionAuthorityBuildersClient, _approveNftCollectionAuthorityClient, approveNftCollectionAuthorityBuilder, approveNftCollectionAuthorityOperation, approveNftCollectionAuthorityOperationHandler };
