// @metaplex-foundation/js/dist/esm/plugins/nftModule/deleteNft.mjs
import { createBurnNftInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { toMintAddress } from './helpers.mjs';
import { findMetadataPda, findMasterEditionV2Pda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TokenProgram } from '../tokenModule/program.mjs';
import { findAssociatedTokenAccountPda } from '../tokenModule/pdas.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _deleteNftClient(nft, input = {}) {
  return this.metaplex.operations().getTask(deleteNftOperation({ ...input,
    mintAddress: toMintAddress(nft)
  }));
}
/** @internal */

function _deleteNftBuildersClient(input) {
  return deleteNftBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'DeleteNftOperation';
const deleteNftOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const deleteNftOperationHandler = {
  handle: async (operation, metaplex) => {
    return deleteNftBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const deleteNftBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    owner = metaplex.identity(),
    ownerTokenAccount,
    collection,
    tokenProgram = TokenProgram.publicKey
  } = params;
  const metadata = findMetadataPda(mintAddress);
  const edition = findMasterEditionV2Pda(mintAddress);
  const tokenAddress = ownerTokenAccount !== null && ownerTokenAccount !== void 0 ? ownerTokenAccount : findAssociatedTokenAccountPda(mintAddress, owner.publicKey);
  return TransactionBuilder.make().add({
    instruction: createBurnNftInstruction({
      metadata,
      owner: owner.publicKey,
      mint: mintAddress,
      tokenAccount: tokenAddress,
      masterEditionAccount: edition,
      splTokenProgram: tokenProgram,
      collectionMetadata: collection ? findMetadataPda(collection) : undefined
    }),
    signers: [owner],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'deleteNft'
  });
};

export { _deleteNftBuildersClient, _deleteNftClient, deleteNftBuilder, deleteNftOperation, deleteNftOperationHandler };
