// @metaplex-foundation/js/dist/esm/plugins/nftModule/freezeDelegatedNft.mjs
import { createFreezeDelegatedAccountInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { toMintAddress } from './helpers.mjs';
import { findMasterEditionV2Pda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TokenProgram } from '../tokenModule/program.mjs';
import { findAssociatedTokenAccountPda } from '../tokenModule/pdas.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _freezeDelegatedNftClient(nft, input) {
  return this.metaplex.operations().getTask(freezeDelegatedNftOperation({ ...input,
    mintAddress: toMintAddress(nft)
  }));
}
/** @internal */

function _freezeDelegatedNftBuildersClient(input) {
  return freezeDelegatedNftBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'FreezeDelegatedNftOperation';
const freezeDelegatedNftOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const freezeDelegatedNftOperationHandler = {
  async handle(operation, metaplex) {
    return freezeDelegatedNftBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }

}; // -----------------
// Builder
// -----------------

const freezeDelegatedNftBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    delegateAuthority,
    tokenOwner = metaplex.identity().publicKey,
    tokenAddress,
    tokenProgram = TokenProgram.publicKey
  } = params;
  const editionAddress = findMasterEditionV2Pda(mintAddress);
  const tokenAddressOrAta = tokenAddress !== null && tokenAddress !== void 0 ? tokenAddress : findAssociatedTokenAccountPda(mintAddress, tokenOwner);
  return TransactionBuilder.make().add({
    instruction: createFreezeDelegatedAccountInstruction({
      delegate: delegateAuthority.publicKey,
      tokenAccount: tokenAddressOrAta,
      edition: editionAddress,
      mint: mintAddress,
      tokenProgram
    }),
    signers: [delegateAuthority],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'freezeDelegatedNft'
  });
};

export { _freezeDelegatedNftBuildersClient, _freezeDelegatedNftClient, freezeDelegatedNftBuilder, freezeDelegatedNftOperation, freezeDelegatedNftOperationHandler };
