// @metaplex-foundation/js/dist/esm/plugins/nftModule/migrateToSizedCollectionNft.mjs
import { createSetCollectionSizeInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { toMintAddress } from './helpers.mjs';
import { findMetadataPda, findCollectionAuthorityRecordPda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _migrateToSizedCollectionNftClient(collection, size, input = {}) {
  return this.metaplex.operations().getTask(migrateToSizedCollectionNftOperation({ ...input,
    mintAddress: toMintAddress(collection),
    size
  }));
}
/** @internal */

function _migrateToSizedCollectionNftBuildersClient(input) {
  return migrateToSizedCollectionNftBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'MigrateToSizedCollectionNftOperation';
const migrateToSizedCollectionNftOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const migrateToSizedCollectionNftOperationHandler = {
  handle: async (operation, metaplex) => {
    return migrateToSizedCollectionNftBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const migrateToSizedCollectionNftBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    collectionAuthority = metaplex.identity(),
    size,
    isDelegated = false
  } = params;
  return TransactionBuilder.make() // Update the metadata account.
  .add({
    instruction: createSetCollectionSizeInstruction({
      collectionMetadata: findMetadataPda(mintAddress),
      collectionAuthority: collectionAuthority.publicKey,
      collectionMint: mintAddress,
      collectionAuthorityRecord: isDelegated ? findCollectionAuthorityRecordPda(mintAddress, collectionAuthority.publicKey) : undefined
    }, {
      setCollectionSizeArgs: {
        size
      }
    }),
    signers: [collectionAuthority],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'setCollectionSize'
  });
};

export { _migrateToSizedCollectionNftBuildersClient, _migrateToSizedCollectionNftClient, migrateToSizedCollectionNftBuilder, migrateToSizedCollectionNftOperation, migrateToSizedCollectionNftOperationHandler };
