// @metaplex-foundation/js/dist/esm/plugins/nftModule/revokeNftCollectionAuthority.mjs
import { createRevokeCollectionAuthorityInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { toMintAddress } from './helpers.mjs';
import { findMetadataPda, findCollectionAuthorityRecordPda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _revokeNftCollectionAuthorityClient(nftOrSft, collectionAuthority, input = {}) {
  return this.metaplex.operations().getTask(revokeNftCollectionAuthorityOperation({ ...input,
    mintAddress: toMintAddress(nftOrSft),
    collectionAuthority
  }));
}
/** @internal */

function _revokeNftCollectionAuthorityBuildersClient(input) {
  return revokeNftCollectionAuthorityBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'RevokeNftCollectionAuthorityOperation';
const revokeNftCollectionAuthorityOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const revokeNftCollectionAuthorityOperationHandler = {
  handle: async (operation, metaplex) => {
    return revokeNftCollectionAuthorityBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const revokeNftCollectionAuthorityBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    collectionAuthority,
    revokeAuthority = metaplex.identity()
  } = params;
  const metadata = findMetadataPda(mintAddress);
  const collectionAuthorityRecord = findCollectionAuthorityRecordPda(mintAddress, collectionAuthority);
  const instruction = createRevokeCollectionAuthorityInstruction({
    collectionAuthorityRecord,
    delegateAuthority: collectionAuthority,
    revokeAuthority: revokeAuthority.publicKey,
    metadata,
    mint: mintAddress
  }); // Temporary fix. The Shank macro wrongfully ask for the delegateAuthority to be a signer.
  // https://github.com/metaplex-foundation/metaplex-program-library/pull/639

  instruction.keys[1].isSigner = false;
  return TransactionBuilder.make() // Revoke the collection authority.
  .add({
    instruction,
    signers: [revokeAuthority],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'revokeCollectionAuthority'
  });
};

export { _revokeNftCollectionAuthorityBuildersClient, _revokeNftCollectionAuthorityClient, revokeNftCollectionAuthorityBuilder, revokeNftCollectionAuthorityOperation, revokeNftCollectionAuthorityOperationHandler };
