// @metaplex-foundation/js/dist/esm/plugins/nftModule/revokeNftUseAuthority.mjs
import { createRevokeUseAuthorityInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { SystemProgram } from '@solana/web3.js';
import { toMintAddress } from './helpers.mjs';
import { findMetadataPda, findUseAuthorityRecordPda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { findAssociatedTokenAccountPda } from '../tokenModule/pdas.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';
import { TokenProgram } from '../tokenModule/program.mjs';

// Clients
// -----------------

/** @internal */

function _revokeNftUseAuthorityClient(nftOrSft, user, input = {}) {
  return this.metaplex.operations().getTask(revokeNftUseAuthorityOperation({ ...input,
    mintAddress: toMintAddress(nftOrSft),
    user
  }));
}
/** @internal */

function _revokeNftUseAuthorityBuildersClient(input) {
  return revokeNftUseAuthorityBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'RevokeNftUseAuthorityOperation';
const revokeNftUseAuthorityOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const revokeNftUseAuthorityOperationHandler = {
  handle: async (operation, metaplex) => {
    return revokeNftUseAuthorityBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const revokeNftUseAuthorityBuilder = (metaplex, params) => {
  var _params$ownerTokenAdd, _params$tokenProgram, _params$systemProgram, _params$instructionKe;

  const {
    mintAddress,
    user,
    owner = metaplex.identity()
  } = params;
  const metadata = findMetadataPda(mintAddress);
  const useAuthorityRecord = findUseAuthorityRecordPda(mintAddress, user);
  const ownerTokenAddress = (_params$ownerTokenAdd = params.ownerTokenAddress) !== null && _params$ownerTokenAdd !== void 0 ? _params$ownerTokenAdd : findAssociatedTokenAccountPda(mintAddress, owner.publicKey);
  return TransactionBuilder.make() // Revoke the use authority.
  .add({
    instruction: createRevokeUseAuthorityInstruction({
      useAuthorityRecord,
      owner: owner.publicKey,
      user,
      ownerTokenAccount: ownerTokenAddress,
      mint: mintAddress,
      metadata,
      tokenProgram: (_params$tokenProgram = params.tokenProgram) !== null && _params$tokenProgram !== void 0 ? _params$tokenProgram : TokenProgram.publicKey,
      systemProgram: (_params$systemProgram = params.systemProgram) !== null && _params$systemProgram !== void 0 ? _params$systemProgram : SystemProgram.programId
    }),
    signers: [owner],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'revokeUseAuthority'
  });
};

export { _revokeNftUseAuthorityBuildersClient, _revokeNftUseAuthorityClient, revokeNftUseAuthorityBuilder, revokeNftUseAuthorityOperation, revokeNftUseAuthorityOperationHandler };
