// @metaplex-foundation/js/dist/esm/plugins/nftModule/useNft.mjs
import { createUtilizeInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { toMintAddress } from './helpers.mjs';
import { findMetadataPda, findUseAuthorityRecordPda, findProgramAsBurnerPda } from './pdas.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { isSigner } from '../../types/Signer.mjs';
import { ExpectedSignerError } from '../../errors/SdkError.mjs';
import { findAssociatedTokenAccountPda } from '../tokenModule/pdas.mjs';
import { toPublicKey } from '../../types/PublicKey.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _useNftClient(nft, input = {}) {
  return this.metaplex.operations().getTask(useNftOperation({ ...input,
    mintAddress: toMintAddress(nft)
  }));
}
/** @internal */

function _useNftBuildersClient(input) {
  return useNftBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'UseNftOperation';
const useNftOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const useNftOperationHandler = {
  handle: async (operation, metaplex) => {
    return useNftBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const useNftBuilder = (metaplex, params) => {
  var _params$ownerTokenAcc, _params$instructionKe;

  const {
    mintAddress,
    numberOfUses = 1,
    owner = metaplex.identity(),
    useAuthority
  } = params;

  if (!isSigner(owner) && !useAuthority) {
    throw new ExpectedSignerError('owner', 'PublicKey', {
      problemSuffix: 'In order to use an NFT you must either provide the owner as a Signer ' + 'or a delegated use authority as a Signer.'
    });
  }

  const metadata = findMetadataPda(mintAddress);
  const tokenAccount = (_params$ownerTokenAcc = params.ownerTokenAccount) !== null && _params$ownerTokenAcc !== void 0 ? _params$ownerTokenAcc : findAssociatedTokenAccountPda(mintAddress, toPublicKey(owner));
  const useAuthorityRecord = useAuthority ? findUseAuthorityRecordPda(mintAddress, useAuthority.publicKey) : undefined;
  const programAsBurner = findProgramAsBurnerPda();
  return TransactionBuilder.make() // Update the metadata account.
  .add({
    instruction: createUtilizeInstruction({
      metadata,
      tokenAccount,
      useAuthority: useAuthority ? useAuthority.publicKey : toPublicKey(owner),
      mint: mintAddress,
      owner: toPublicKey(owner),
      useAuthorityRecord,
      burner: useAuthorityRecord ? programAsBurner : undefined
    }, {
      utilizeArgs: {
        numberOfUses
      }
    }),
    signers: [owner, useAuthority].filter(isSigner),
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'utilizeNft'
  });
};

export { _useNftBuildersClient, _useNftClient, useNftBuilder, useNftOperation, useNftOperationHandler };
