// @metaplex-foundation/js/dist/esm/plugins/nftModule/verifyNftCreator.mjs
import { createSignMetadataInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { findMetadataPda } from './pdas.mjs';
import { toMintAddress } from './helpers.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _verifyNftCreatorClient(nftOrSft, creator, input = {}) {
  return this.metaplex.operations().getTask(verifyNftCreatorOperation({ ...input,
    mintAddress: toMintAddress(nftOrSft),
    creator
  }));
}
/** @internal */

function _verifyNftCreatorBuildersClient(input) {
  return verifyNftCreatorBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'VerifyNftCreatorOperation';
const verifyNftCreatorOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const verifyNftCreatorOperationHandler = {
  handle: async (operation, metaplex) => {
    return verifyNftCreatorBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const verifyNftCreatorBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    creator = metaplex.identity()
  } = params;
  return TransactionBuilder.make() // Verify the creator.
  .add({
    instruction: createSignMetadataInstruction({
      metadata: findMetadataPda(mintAddress),
      creator: creator.publicKey
    }),
    signers: [creator],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'verifyCreator'
  });
};

export { _verifyNftCreatorBuildersClient, _verifyNftCreatorClient, verifyNftCreatorBuilder, verifyNftCreatorOperation, verifyNftCreatorOperationHandler };
