// @metaplex-foundation/js/dist/esm/plugins/nftModule/verifyNftCollection.mjs
import { createVerifySizedCollectionItemInstruction, createVerifyCollectionInstruction } from '@metaplex-foundation/mpl-token-metadata';
import { toMetadataAccount } from './accounts.mjs';
import { ParentCollectionMissingError } from './errors.mjs';
import { toMintAddress } from './helpers.mjs';
import { toMetadata } from './Metadata.mjs';
import { findMetadataPda, findMasterEditionV2Pda, findCollectionAuthorityRecordPda } from './pdas.mjs';
import { Task } from '../../utils/Task.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// Clients
// -----------------

/** @internal */

function _verifyNftCollectionClient(nftOrSft, input = {}) {
  return new Task(async scope => {
    var _input$collectionMint;

    const mintAddress = toMintAddress(nftOrSft);
    const collectionFromNft = 'collection' in nftOrSft && nftOrSft.collection ? nftOrSft.collection.address : undefined;
    let collectionMintAddress = (_input$collectionMint = input.collectionMintAddress) !== null && _input$collectionMint !== void 0 ? _input$collectionMint : collectionFromNft;

    if (!('collection' in nftOrSft) && !collectionMintAddress) {
      const metadataAddress = findMetadataPda(mintAddress);
      const metadata = toMetadata(toMetadataAccount(await this.metaplex.rpc().getAccount(metadataAddress)));
      scope.throwIfCanceled();
      collectionMintAddress = metadata.collection ? metadata.collection.address : undefined;
    }

    if (!collectionMintAddress) {
      throw new ParentCollectionMissingError(mintAddress, 'verifyCollection');
    }

    return this.metaplex.operations().execute(verifyNftCollectionOperation({ ...input,
      mintAddress,
      collectionMintAddress
    }), scope);
  });
}
/** @internal */

function _verifyNftCollectionBuildersClient(input) {
  return verifyNftCollectionBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'VerifyNftCollectionOperation';
const verifyNftCollectionOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const verifyNftCollectionOperationHandler = {
  handle: async (operation, metaplex) => {
    return verifyNftCollectionBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const verifyNftCollectionBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    collectionMintAddress,
    isSizedCollection = true,
    isDelegated = false,
    collectionAuthority = metaplex.identity(),
    payer = metaplex.identity()
  } = params;
  const accounts = {
    metadata: findMetadataPda(mintAddress),
    collectionAuthority: collectionAuthority.publicKey,
    payer: payer.publicKey,
    collectionMint: collectionMintAddress,
    collection: findMetadataPda(collectionMintAddress),
    collectionMasterEditionAccount: findMasterEditionV2Pda(collectionMintAddress)
  };
  const instruction = isSizedCollection ? createVerifySizedCollectionItemInstruction(accounts) : createVerifyCollectionInstruction(accounts);

  if (isDelegated) {
    instruction.keys.push({
      pubkey: findCollectionAuthorityRecordPda(collectionMintAddress, collectionAuthority.publicKey),
      isWritable: false,
      isSigner: false
    });
  }

  return TransactionBuilder.make().setFeePayer(payer) // Verify the collection.
  .add({
    instruction: instruction,
    signers: [payer, collectionAuthority],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'verifyCollection'
  });
};

export { _verifyNftCollectionBuildersClient, _verifyNftCollectionClient, verifyNftCollectionBuilder, verifyNftCollectionOperation, verifyNftCollectionOperationHandler };
