// @metaplex-foundation/js/dist/esm/plugins/auctionHouseModule/accounts.mjs
import { Auctioneer, AuctionHouse, ListingReceipt, BidReceipt, PurchaseReceipt } from '@metaplex-foundation/mpl-auction-house';
import { getAccountParsingFunction, getAccountParsingAndAssertingFunction } from '../../types/Account.mjs';

const parseAuctioneerAccount = getAccountParsingFunction(Auctioneer);
const toAuctioneerAccount = getAccountParsingAndAssertingFunction(Auctioneer);
const parseAuctionHouseAccount = getAccountParsingFunction(AuctionHouse);
const toAuctionHouseAccount = getAccountParsingAndAssertingFunction(AuctionHouse);
const parseListingReceiptAccount = getAccountParsingFunction(ListingReceipt);
const toListingReceiptAccount = getAccountParsingAndAssertingFunction(ListingReceipt);
const parseBidReceiptAccount = getAccountParsingFunction(BidReceipt);
const toBidReceiptAccount = getAccountParsingAndAssertingFunction(BidReceipt);
const parsePurchaseReceiptAccount = getAccountParsingFunction(PurchaseReceipt);
const toPurchaseReceiptAccount = getAccountParsingAndAssertingFunction(PurchaseReceipt);

export { parseAuctionHouseAccount, parseAuctioneerAccount, parseBidReceiptAccount, parseListingReceiptAccount, parsePurchaseReceiptAccount, toAuctionHouseAccount, toAuctioneerAccount, toBidReceiptAccount, toListingReceiptAccount, toPurchaseReceiptAccount };
