// @metaplex-foundation/js/dist/esm/plugins/auctionHouseModule/loadBid.mjs
import assert from '../../utils/assert.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { assertNftOrSftWithToken } from '../nftModule/Nft.mjs';
import { amount } from '../../types/Amount.mjs';

// Operation
// -----------------

const Key = 'LoadBidOperation';
const loadBidOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const loadBidOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const {
      lazyBid,
      loadJsonMetadata = true,
      commitment
    } = operation.input;
    const bid = { ...lazyBid,
      model: 'bid',
      lazy: false
    };

    if (lazyBid.tokenAddress) {
      const asset = await metaplex.nfts().findByToken(lazyBid.tokenAddress, {
        commitment,
        loadJsonMetadata
      }).run(scope);
      scope.throwIfCanceled();
      assertNftOrSftWithToken(asset);
      assert(asset.metadataAddress.equals(lazyBid.metadataAddress), `Asset metadata address must be ${lazyBid.metadataAddress}`);
      return { ...bid,
        isPublic: false,
        asset,
        tokens: amount(lazyBid.tokens, asset.mint.currency)
      };
    } else {
      const asset = await metaplex.nfts().findByMetadata(lazyBid.metadataAddress, {
        commitment,
        loadJsonMetadata
      }).run(scope);
      scope.throwIfCanceled();
      return { ...bid,
        isPublic: true,
        asset,
        tokens: amount(lazyBid.tokens, asset.mint.currency)
      };
    }
  }
};

export { loadBidOperation, loadBidOperationHandler };
