// @metaplex-foundation/js/dist/cjs/plugins/nftModule/approveNftCollectionAuthority.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplTokenMetadata = require('@metaplex-foundation/mpl-token-metadata');
var web3_js = require('@solana/web3.js');
var helpers = require('./helpers.cjs');
var pdas = require('./pdas.cjs');
var Operation = require('../../types/Operation.cjs');
var TransactionBuilder = require('../../utils/TransactionBuilder.cjs');

// Clients
// -----------------

/** @internal */

function _approveNftCollectionAuthorityClient(nftOrSft, collectionAuthority, input = {}) {
  return this.metaplex.operations().getTask(approveNftCollectionAuthorityOperation({ ...input,
    mintAddress: helpers.toMintAddress(nftOrSft),
    collectionAuthority
  }));
}
/** @internal */

function _approveNftCollectionAuthorityBuildersClient(input) {
  return approveNftCollectionAuthorityBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'ApproveNftCollectionAuthorityOperation';
const approveNftCollectionAuthorityOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const approveNftCollectionAuthorityOperationHandler = {
  handle: async (operation, metaplex) => {
    return approveNftCollectionAuthorityBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const approveNftCollectionAuthorityBuilder = (metaplex, params) => {
  var _params$systemProgram, _params$instructionKe;

  const {
    mintAddress,
    collectionAuthority,
    updateAuthority = metaplex.identity(),
    payer = metaplex.identity()
  } = params;
  const metadata = pdas.findMetadataPda(mintAddress);
  const collectionAuthorityRecord = pdas.findCollectionAuthorityRecordPda(mintAddress, collectionAuthority);
  return TransactionBuilder.TransactionBuilder.make().setFeePayer(payer) // Approve the collection authority.
  .add({
    instruction: mplTokenMetadata.createApproveCollectionAuthorityInstruction({
      collectionAuthorityRecord,
      newCollectionAuthority: collectionAuthority,
      updateAuthority: updateAuthority.publicKey,
      payer: payer.publicKey,
      metadata,
      mint: mintAddress,
      systemProgram: (_params$systemProgram = params.systemProgram) !== null && _params$systemProgram !== void 0 ? _params$systemProgram : web3_js.SystemProgram.programId
    }),
    signers: [payer, updateAuthority],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'approveCollectionAuthority'
  });
};

exports._approveNftCollectionAuthorityBuildersClient = _approveNftCollectionAuthorityBuildersClient;
exports._approveNftCollectionAuthorityClient = _approveNftCollectionAuthorityClient;
exports.approveNftCollectionAuthorityBuilder = approveNftCollectionAuthorityBuilder;
exports.approveNftCollectionAuthorityOperation = approveNftCollectionAuthorityOperation;
exports.approveNftCollectionAuthorityOperationHandler = approveNftCollectionAuthorityOperationHandler;
