// @metaplex-foundation/js/dist/cjs/plugins/nftModule/createSft.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplTokenMetadata = require('@metaplex-foundation/mpl-token-metadata');
var web3_js = require('@solana/web3.js');
var pdas = require('./pdas.cjs');
var Sft = require('./Sft.cjs');
var Task = require('../../utils/Task.cjs');
var Operation = require('../../types/Operation.cjs');
var TransactionBuilder = require('../../utils/TransactionBuilder.cjs');
var pdas$1 = require('../tokenModule/pdas.cjs');
var PublicKey = require('../../types/PublicKey.cjs');
var Signer = require('../../types/Signer.cjs');

// Clients
// -----------------

/** @internal */

function _createSftClient(input) {
  return new Task.Task(async scope => {
    var _output$tokenAddress;

    const operation = createSftOperation(input);
    const output = await this.metaplex.operations().execute(operation, scope);
    scope.throwIfCanceled();
    const sft = await this.findByMint(output.mintAddress, {
      tokenAddress: (_output$tokenAddress = output.tokenAddress) !== null && _output$tokenAddress !== void 0 ? _output$tokenAddress : undefined
    }).run(scope);
    Sft.assertSft(sft);
    return { ...output,
      sft
    };
  });
}
/** @internal */

function _createSftBuildersClient(input) {
  return createSftBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'CreateSftOperation';
const createSftOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const createSftOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const builder = await createSftBuilder(metaplex, operation.input);
    scope.throwIfCanceled();
    return builder.sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const createSftBuilder = async (metaplex, params) => {
  var _params$creators, _params$symbol, _params$uses, _params$isMutable, _params$createMetadat;

  const {
    payer = metaplex.identity(),
    useNewMint = web3_js.Keypair.generate(),
    updateAuthority = metaplex.identity(),
    mintAuthority = metaplex.identity()
  } = params;
  const mintAndTokenBuilder = await createMintAndTokenForSftBuilder(metaplex, params, useNewMint);
  const {
    mintAddress,
    tokenAddress
  } = mintAndTokenBuilder.getContext();
  const metadataPda = pdas.findMetadataPda(mintAddress);
  const creatorsInput = (_params$creators = params.creators) !== null && _params$creators !== void 0 ? _params$creators : [{
    address: updateAuthority.publicKey,
    authority: updateAuthority,
    share: 100
  }];
  const creators = creatorsInput.length > 0 ? creatorsInput.map(creator => ({ ...creator,
    verified: creator.address.equals(updateAuthority.publicKey)
  })) : null;
  const createMetadataInstruction = mplTokenMetadata.createCreateMetadataAccountV3Instruction({
    metadata: metadataPda,
    mint: mintAddress,
    mintAuthority: mintAuthority.publicKey,
    payer: payer.publicKey,
    updateAuthority: updateAuthority.publicKey
  }, {
    createMetadataAccountArgsV3: {
      data: {
        name: params.name,
        symbol: (_params$symbol = params.symbol) !== null && _params$symbol !== void 0 ? _params$symbol : '',
        uri: params.uri,
        sellerFeeBasisPoints: params.sellerFeeBasisPoints,
        creators,
        collection: params.collection ? {
          key: params.collection,
          verified: false
        } : null,
        uses: (_params$uses = params.uses) !== null && _params$uses !== void 0 ? _params$uses : null
      },
      isMutable: (_params$isMutable = params.isMutable) !== null && _params$isMutable !== void 0 ? _params$isMutable : true,
      collectionDetails: params.isCollection ? {
        __kind: 'V1',
        size: 0
      } // Program will hardcode size to zero anyway.
      : null
    }
  }); // When the payer is different than the update authority, the latter will
  // not be marked as a signer and therefore signing as a creator will fail.

  createMetadataInstruction.keys[4].isSigner = true;
  const verifyAdditionalCreatorInstructions = creatorsInput.filter(creator => {
    return !!creator.authority && !creator.address.equals(updateAuthority.publicKey);
  }).map(creator => {
    return metaplex.nfts().builders().verifyCreator({
      mintAddress,
      creator: creator.authority
    });
  });
  return TransactionBuilder.TransactionBuilder.make().setFeePayer(payer).setContext({
    mintAddress,
    metadataAddress: metadataPda,
    tokenAddress
  }) // Create the mint and token accounts before minting 1 token to the owner.
  .add(mintAndTokenBuilder) // Create metadata account.
  .add({
    instruction: createMetadataInstruction,
    signers: [payer, mintAuthority, updateAuthority],
    key: (_params$createMetadat = params.createMetadataInstructionKey) !== null && _params$createMetadat !== void 0 ? _params$createMetadat : 'createMetadata'
  }) // Verify additional creators.
  .add(...verifyAdditionalCreatorInstructions) // Verify collection.
  .when(!!params.collection && !!params.collectionAuthority, builder => {
    var _params$collectionAut, _params$collectionIsS;

    return builder.add(metaplex.nfts().builders().verifyCollection({
      payer,
      mintAddress,
      collectionMintAddress: params.collection,
      collectionAuthority: params.collectionAuthority,
      isDelegated: (_params$collectionAut = params.collectionAuthorityIsDelegated) !== null && _params$collectionAut !== void 0 ? _params$collectionAut : false,
      isSizedCollection: (_params$collectionIsS = params.collectionIsSized) !== null && _params$collectionIsS !== void 0 ? _params$collectionIsS : true
    }));
  });
};

const createMintAndTokenForSftBuilder = async (metaplex, params, useNewMint) => {
  var _params$useExistingMi;

  const {
    payer = metaplex.identity(),
    mintAuthority = metaplex.identity(),
    freezeAuthority = metaplex.identity().publicKey
  } = params;
  const mintAddress = (_params$useExistingMi = params.useExistingMint) !== null && _params$useExistingMi !== void 0 ? _params$useExistingMi : useNewMint.publicKey;
  const associatedTokenAddress = params.tokenOwner ? pdas$1.findAssociatedTokenAccountPda(mintAddress, params.tokenOwner) : null;
  const tokenAddress = params.tokenAddress ? PublicKey.toPublicKey(params.tokenAddress) : associatedTokenAddress;
  let tokenExists;

  if (!!params.useExistingMint && !!tokenAddress) {
    const tokenAccount = await metaplex.rpc().getAccount(tokenAddress);
    tokenExists = tokenAccount.exists;
  } else {
    tokenExists = false;
  }

  const builder = TransactionBuilder.TransactionBuilder.make().setContext({
    mintAddress,
    tokenAddress
  }); // Create the mint account if it doesn't exist.

  if (!params.useExistingMint) {
    var _params$decimals;

    builder.add(await metaplex.tokens().builders().createMint({
      decimals: (_params$decimals = params.decimals) !== null && _params$decimals !== void 0 ? _params$decimals : 0,
      mint: useNewMint,
      payer,
      mintAuthority: mintAuthority.publicKey,
      freezeAuthority,
      tokenProgram: params.tokenProgram,
      createAccountInstructionKey: params.createMintAccountInstructionKey,
      initializeMintInstructionKey: params.initializeMintInstructionKey
    }));
  } // Create the token account if it doesn't exist.


  const isNewToken = !!params.tokenAddress && Signer.isSigner(params.tokenAddress);
  const isNewAssociatedToken = !!params.tokenOwner;

  if (!tokenExists && (isNewToken || isNewAssociatedToken)) {
    builder.add(await metaplex.tokens().builders().createToken({
      mint: mintAddress,
      owner: params.tokenOwner,
      token: params.tokenAddress,
      payer,
      tokenProgram: params.tokenProgram,
      associatedTokenProgram: params.associatedTokenProgram,
      createAssociatedTokenAccountInstructionKey: params.createAssociatedTokenAccountInstructionKey,
      createAccountInstructionKey: params.createTokenAccountInstructionKey,
      initializeTokenInstructionKey: params.initializeTokenInstructionKey
    }));
  } // Mint provided amount to the token account.


  if (tokenAddress && params.tokenAmount) {
    builder.add(await metaplex.tokens().builders().mint({
      mint: mintAddress,
      toToken: tokenAddress,
      toTokenExists: true,
      amount: params.tokenAmount,
      mintAuthority,
      tokenProgram: params.tokenProgram,
      mintTokensInstructionKey: params.mintTokensInstructionKey
    }));
  }

  return builder;
};

exports._createSftBuildersClient = _createSftBuildersClient;
exports._createSftClient = _createSftClient;
exports.createSftBuilder = createSftBuilder;
exports.createSftOperation = createSftOperation;
exports.createSftOperationHandler = createSftOperationHandler;
