// @metaplex-foundation/js/dist/cjs/plugins/nftModule/deleteNft.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplTokenMetadata = require('@metaplex-foundation/mpl-token-metadata');
var helpers = require('./helpers.cjs');
var pdas = require('./pdas.cjs');
var Operation = require('../../types/Operation.cjs');
var program = require('../tokenModule/program.cjs');
var pdas$1 = require('../tokenModule/pdas.cjs');
var TransactionBuilder = require('../../utils/TransactionBuilder.cjs');

// Clients
// -----------------

/** @internal */

function _deleteNftClient(nft, input = {}) {
  return this.metaplex.operations().getTask(deleteNftOperation({ ...input,
    mintAddress: helpers.toMintAddress(nft)
  }));
}
/** @internal */

function _deleteNftBuildersClient(input) {
  return deleteNftBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'DeleteNftOperation';
const deleteNftOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const deleteNftOperationHandler = {
  handle: async (operation, metaplex) => {
    return deleteNftBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const deleteNftBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    owner = metaplex.identity(),
    ownerTokenAccount,
    collection,
    tokenProgram = program.TokenProgram.publicKey
  } = params;
  const metadata = pdas.findMetadataPda(mintAddress);
  const edition = pdas.findMasterEditionV2Pda(mintAddress);
  const tokenAddress = ownerTokenAccount !== null && ownerTokenAccount !== void 0 ? ownerTokenAccount : pdas$1.findAssociatedTokenAccountPda(mintAddress, owner.publicKey);
  return TransactionBuilder.TransactionBuilder.make().add({
    instruction: mplTokenMetadata.createBurnNftInstruction({
      metadata,
      owner: owner.publicKey,
      mint: mintAddress,
      tokenAccount: tokenAddress,
      masterEditionAccount: edition,
      splTokenProgram: tokenProgram,
      collectionMetadata: collection ? pdas.findMetadataPda(collection) : undefined
    }),
    signers: [owner],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'deleteNft'
  });
};

exports._deleteNftBuildersClient = _deleteNftBuildersClient;
exports._deleteNftClient = _deleteNftClient;
exports.deleteNftBuilder = deleteNftBuilder;
exports.deleteNftOperation = deleteNftOperation;
exports.deleteNftOperationHandler = deleteNftOperationHandler;
