// @metaplex-foundation/js/dist/cjs/plugins/nftModule/loadMetadata.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Operation = require('../../types/Operation.cjs');

// -----------------
// Clients
// -----------------

/** @internal */
function _loadMetadataClient(metadata, options) {
  return this.metaplex.operations().getTask(loadMetadataOperation({
    metadata,
    ...options
  }));
} // -----------------
// Operation
// -----------------

const Key = 'LoadMetadataOperation';
const loadMetadataOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const loadMetadataOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const {
      metadata,
      loadJsonMetadata = true
    } = operation.input;
    let nftOrSft = await metaplex.nfts().findByMint(metadata.mintAddress, { ...operation.input,
      loadJsonMetadata: !metadata.jsonLoaded && loadJsonMetadata
    }).run(scope);

    if (!nftOrSft.jsonLoaded && metadata.jsonLoaded) {
      nftOrSft = { ...nftOrSft,
        json: metadata.json,
        jsonLoaded: true
      };
    }

    return nftOrSft;
  }
};

exports._loadMetadataClient = _loadMetadataClient;
exports.loadMetadataOperation = loadMetadataOperation;
exports.loadMetadataOperationHandler = loadMetadataOperationHandler;
