// @metaplex-foundation/js/dist/cjs/plugins/nftModule/Nft.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Sft = require('./Sft.cjs');
var assert = require('../../utils/assert.cjs');

const isNft = value => typeof value === 'object' && value.model === 'nft';
function assertNft(value) {
  assert["default"](isNft(value), `Expected Nft model`);
}
const toNft = (metadata, mint, edition) => ({ ...Sft.toSft(metadata, mint),
  model: 'nft',
  edition
});
const isNftWithToken = value => isNft(value) && 'token' in value;
function assertNftWithToken(value) {
  assert["default"](isNftWithToken(value), `Expected Nft model with token`);
}
function assertNftOrSftWithToken(value) {
  assert["default"](isNftWithToken(value) || Sft.isSftWithToken(value), `Expected Nft or Sft model with token`);
}
const toNftWithToken = (metadata, mint, edition, token) => ({ ...Sft.toSftWithToken(metadata, mint, token),
  model: 'nft',
  edition
});

exports.assertNft = assertNft;
exports.assertNftOrSftWithToken = assertNftOrSftWithToken;
exports.assertNftWithToken = assertNftWithToken;
exports.isNft = isNft;
exports.isNftWithToken = isNftWithToken;
exports.toNft = toNft;
exports.toNftWithToken = toNftWithToken;
