// @metaplex-foundation/js/dist/cjs/plugins/nftModule/plugin.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplTokenMetadata = require('@metaplex-foundation/mpl-token-metadata');
var approveNftCollectionAuthority = require('./approveNftCollectionAuthority.cjs');
var approveNftUseAuthority = require('./approveNftUseAuthority.cjs');
var createNft = require('./createNft.cjs');
var createSft = require('./createSft.cjs');
var deleteNft = require('./deleteNft.cjs');
var findNftByMetadata = require('./findNftByMetadata.cjs');
var findNftByMint = require('./findNftByMint.cjs');
var findNftByToken = require('./findNftByToken.cjs');
var findNftsByCreator = require('./findNftsByCreator.cjs');
var findNftsByMintList = require('./findNftsByMintList.cjs');
var findNftsByOwner = require('./findNftsByOwner.cjs');
var findNftsByUpdateAuthority = require('./findNftsByUpdateAuthority.cjs');
var freezeDelegatedNft = require('./freezeDelegatedNft.cjs');
var gpaBuilders = require('./gpaBuilders.cjs');
var loadMetadata = require('./loadMetadata.cjs');
var migrateToSizedCollectionNft = require('./migrateToSizedCollectionNft.cjs');
var NftClient = require('./NftClient.cjs');
var printNewEdition = require('./printNewEdition.cjs');
var program = require('./program.cjs');
var revokeNftCollectionAuthority = require('./revokeNftCollectionAuthority.cjs');
var revokeNftUseAuthority = require('./revokeNftUseAuthority.cjs');
var thawDelegatedNft = require('./thawDelegatedNft.cjs');
var unverifyNftCollection = require('./unverifyNftCollection.cjs');
var unverifyNftCreator = require('./unverifyNftCreator.cjs');
var updateNft = require('./updateNft.cjs');
var uploadMetadata = require('./uploadMetadata.cjs');
var useNft = require('./useNft.cjs');
var verifyNftCollection = require('./verifyNftCollection.cjs');
var verifyNftCreator = require('./verifyNftCreator.cjs');

const nftModule = () => ({
  install(metaplex) {
    // Token Metadata Program.
    metaplex.programs().register({
      name: 'TokenMetadataProgram',
      address: program.TokenMetadataProgram.publicKey,
      errorResolver: error => mplTokenMetadata.cusper.errorFromProgramLogs(error.logs, false),
      gpaResolver: metaplex => new gpaBuilders.TokenMetadataGpaBuilder(metaplex, program.TokenMetadataProgram.publicKey)
    }); // Operations.

    const op = metaplex.operations();
    op.register(approveNftCollectionAuthority.approveNftCollectionAuthorityOperation, approveNftCollectionAuthority.approveNftCollectionAuthorityOperationHandler);
    op.register(approveNftUseAuthority.approveNftUseAuthorityOperation, approveNftUseAuthority.approveNftUseAuthorityOperationHandler);
    op.register(createNft.createNftOperation, createNft.createNftOperationHandler);
    op.register(createSft.createSftOperation, createSft.createSftOperationHandler);
    op.register(deleteNft.deleteNftOperation, deleteNft.deleteNftOperationHandler);
    op.register(findNftByMetadata.findNftByMetadataOperation, findNftByMetadata.findNftByMetadataOperationHandler);
    op.register(findNftByMint.findNftByMintOperation, findNftByMint.findNftByMintOperationHandler);
    op.register(findNftByToken.findNftByTokenOperation, findNftByToken.findNftByTokenOperationHandler);
    op.register(findNftsByCreator.findNftsByCreatorOperation, findNftsByCreator.findNftsByCreatorOperationHandler);
    op.register(findNftsByMintList.findNftsByMintListOperation, findNftsByMintList.findNftsByMintListOperationHandler);
    op.register(findNftsByOwner.findNftsByOwnerOperation, findNftsByOwner.findNftsByOwnerOperationHandler);
    op.register(findNftsByUpdateAuthority.findNftsByUpdateAuthorityOperation, findNftsByUpdateAuthority.findNftsByUpdateAuthorityOperationHandler);
    op.register(freezeDelegatedNft.freezeDelegatedNftOperation, freezeDelegatedNft.freezeDelegatedNftOperationHandler);
    op.register(loadMetadata.loadMetadataOperation, loadMetadata.loadMetadataOperationHandler);
    op.register(migrateToSizedCollectionNft.migrateToSizedCollectionNftOperation, migrateToSizedCollectionNft.migrateToSizedCollectionNftOperationHandler);
    op.register(printNewEdition.printNewEditionOperation, printNewEdition.printNewEditionOperationHandler);
    op.register(revokeNftCollectionAuthority.revokeNftCollectionAuthorityOperation, revokeNftCollectionAuthority.revokeNftCollectionAuthorityOperationHandler);
    op.register(revokeNftUseAuthority.revokeNftUseAuthorityOperation, revokeNftUseAuthority.revokeNftUseAuthorityOperationHandler);
    op.register(thawDelegatedNft.thawDelegatedNftOperation, thawDelegatedNft.thawDelegatedNftOperationHandler);
    op.register(unverifyNftCollection.unverifyNftCollectionOperation, unverifyNftCollection.unverifyNftCollectionOperationHandler);
    op.register(unverifyNftCreator.unverifyNftCreatorOperation, unverifyNftCreator.unverifyNftCreatorOperationHandler);
    op.register(updateNft.updateNftOperation, updateNft.updateNftOperationHandler);
    op.register(uploadMetadata.uploadMetadataOperation, uploadMetadata.uploadMetadataOperationHandler);
    op.register(useNft.useNftOperation, useNft.useNftOperationHandler);
    op.register(verifyNftCollection.verifyNftCollectionOperation, verifyNftCollection.verifyNftCollectionOperationHandler);
    op.register(verifyNftCreator.verifyNftCreatorOperation, verifyNftCreator.verifyNftCreatorOperationHandler);

    metaplex.nfts = function () {
      return new NftClient.NftClient(this);
    };
  }

});

exports.nftModule = nftModule;
