// @metaplex-foundation/js/dist/cjs/plugins/nftModule/verifyNftCreator.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplTokenMetadata = require('@metaplex-foundation/mpl-token-metadata');
var pdas = require('./pdas.cjs');
var helpers = require('./helpers.cjs');
var Operation = require('../../types/Operation.cjs');
var TransactionBuilder = require('../../utils/TransactionBuilder.cjs');

// Clients
// -----------------

/** @internal */

function _verifyNftCreatorClient(nftOrSft, creator, input = {}) {
  return this.metaplex.operations().getTask(verifyNftCreatorOperation({ ...input,
    mintAddress: helpers.toMintAddress(nftOrSft),
    creator
  }));
}
/** @internal */

function _verifyNftCreatorBuildersClient(input) {
  return verifyNftCreatorBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'VerifyNftCreatorOperation';
const verifyNftCreatorOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const verifyNftCreatorOperationHandler = {
  handle: async (operation, metaplex) => {
    return verifyNftCreatorBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const verifyNftCreatorBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    creator = metaplex.identity()
  } = params;
  return TransactionBuilder.TransactionBuilder.make() // Verify the creator.
  .add({
    instruction: mplTokenMetadata.createSignMetadataInstruction({
      metadata: pdas.findMetadataPda(mintAddress),
      creator: creator.publicKey
    }),
    signers: [creator],
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'verifyCreator'
  });
};

exports._verifyNftCreatorBuildersClient = _verifyNftCreatorBuildersClient;
exports._verifyNftCreatorClient = _verifyNftCreatorClient;
exports.verifyNftCreatorBuilder = verifyNftCreatorBuilder;
exports.verifyNftCreatorOperation = verifyNftCreatorOperation;
exports.verifyNftCreatorOperationHandler = verifyNftCreatorOperationHandler;
