// @metaplex-foundation/js/dist/cjs/plugins/candyMachineModule/CandyMachinesBuildersClient.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var createCandyMachine = require('./createCandyMachine.cjs');
var deleteCandyMachine = require('./deleteCandyMachine.cjs');
var insertItemsToCandyMachine = require('./insertItemsToCandyMachine.cjs');
var mintCandyMachine = require('./mintCandyMachine.cjs');
var updateCandyMachine = require('./updateCandyMachine.cjs');

class CandyMachinesBuildersClient {
  constructor(metaplex) {
    this.metaplex = metaplex;
  }

  create(input) {
    return createCandyMachine.createCandyMachineBuilder(this.metaplex, input);
  }

  delete(input) {
    return deleteCandyMachine.deleteCandyMachineBuilder(this.metaplex, input);
  }

  insertItems(input) {
    return insertItemsToCandyMachine.insertItemsToCandyMachineBuilder(input);
  }

  mint(input) {
    return mintCandyMachine.mintCandyMachineBuilder(this.metaplex, input);
  }

  update(input) {
    return updateCandyMachine.updateCandyMachineBuilder(this.metaplex, input);
  }

}

exports.CandyMachinesBuildersClient = CandyMachinesBuildersClient;
