// @metaplex-foundation/js/dist/esm/plugins/tokenModule/freezeTokens.mjs
import { createFreezeAccountInstruction } from '@solana/spl-token';
import { findAssociatedTokenAccountPda } from './pdas.mjs';
import { TokenProgram } from './program.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { isSigner } from '../../types/Signer.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// -----------------
// Clients
// -----------------

/** @internal */
function _freezeTokensClient(input) {
  return this.metaplex.operations().getTask(freezeTokensOperation(input));
}
/** @internal */

function _freezeTokensBuildersClient(input) {
  return freezeTokensBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'FreezeTokensOperation';
const freezeTokensOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const freezeTokensOperationHandler = {
  async handle(operation, metaplex) {
    return freezeTokensBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }

}; // -----------------
// Builder
// -----------------

const freezeTokensBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    tokenOwner = metaplex.identity().publicKey,
    tokenAddress,
    multiSigners = [],
    freezeAuthority,
    tokenProgram = TokenProgram.publicKey
  } = params;
  const [authorityPublicKey, signers] = isSigner(freezeAuthority) ? [freezeAuthority.publicKey, [freezeAuthority]] : [freezeAuthority, multiSigners];
  const tokenAddressOrAta = tokenAddress !== null && tokenAddress !== void 0 ? tokenAddress : findAssociatedTokenAccountPda(mintAddress, tokenOwner);
  return TransactionBuilder.make().add({
    instruction: createFreezeAccountInstruction(tokenAddressOrAta, mintAddress, authorityPublicKey, multiSigners, tokenProgram),
    signers,
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'freezeTokens'
  });
};

export { _freezeTokensBuildersClient, _freezeTokensClient, freezeTokensBuilder, freezeTokensOperation, freezeTokensOperationHandler };
