// @metaplex-foundation/js/dist/esm/plugins/tokenModule/revokeTokenDelegateAuthority.mjs
import { createRevokeInstruction } from '@solana/spl-token';
import { findAssociatedTokenAccountPda } from './pdas.mjs';
import { TokenProgram } from './program.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { isSigner } from '../../types/Signer.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// -----------------
// Clients
// -----------------

/** @internal */
function _revokeTokenDelegateAuthorityClient(input) {
  return this.metaplex.operations().getTask(revokeTokenDelegateAuthorityOperation(input));
}
/** @internal */

function _revokeTokenDelegateAuthorityBuildersClient(input) {
  return revokeTokenDelegateAuthorityBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'RevokeTokenDelegateAuthorityOperation';
const revokeTokenDelegateAuthorityOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const revokeTokenDelegateAuthorityOperationHandler = {
  handle: async (operation, metaplex) => {
    return revokeTokenDelegateAuthorityBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }
}; // -----------------
// Builder
// -----------------

const revokeTokenDelegateAuthorityBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    owner = metaplex.identity(),
    tokenAddress,
    multiSigners = [],
    tokenProgram = TokenProgram.publicKey
  } = params;
  const [ownerPublicKey, signers] = isSigner(owner) ? [owner.publicKey, [owner]] : [owner, multiSigners];
  const tokenAccount = tokenAddress !== null && tokenAddress !== void 0 ? tokenAddress : findAssociatedTokenAccountPda(mintAddress, ownerPublicKey);
  return TransactionBuilder.make().add({
    instruction: createRevokeInstruction(tokenAccount, ownerPublicKey, multiSigners, tokenProgram),
    signers,
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'revokeDelegateAuthority'
  });
};

export { _revokeTokenDelegateAuthorityBuildersClient, _revokeTokenDelegateAuthorityClient, revokeTokenDelegateAuthorityBuilder, revokeTokenDelegateAuthorityOperation, revokeTokenDelegateAuthorityOperationHandler };
