// @metaplex-foundation/js/dist/esm/plugins/tokenModule/thawTokens.mjs
import { createThawAccountInstruction } from '@solana/spl-token';
import { findAssociatedTokenAccountPda } from './pdas.mjs';
import { TokenProgram } from './program.mjs';
import { useOperation } from '../../types/Operation.mjs';
import { isSigner } from '../../types/Signer.mjs';
import { TransactionBuilder } from '../../utils/TransactionBuilder.mjs';

// -----------------
// Clients
// -----------------

/** @internal */
function _thawTokensClient(input) {
  return this.metaplex.operations().getTask(thawTokensOperation(input));
}
/** @internal */

function _thawTokensBuildersClient(input) {
  return thawTokensBuilder(this.metaplex, input);
} // -----------------
// Operation
// -----------------

const Key = 'ThawTokensOperation';
const thawTokensOperation = useOperation(Key);
// -----------------
// Handler
// -----------------
const thawTokensOperationHandler = {
  async handle(operation, metaplex) {
    return thawTokensBuilder(metaplex, operation.input).sendAndConfirm(metaplex, operation.input.confirmOptions);
  }

}; // -----------------
// Builder
// -----------------

const thawTokensBuilder = (metaplex, params) => {
  var _params$instructionKe;

  const {
    mintAddress,
    tokenOwner = metaplex.identity().publicKey,
    tokenAddress,
    multiSigners = [],
    freezeAuthority,
    tokenProgram = TokenProgram.publicKey
  } = params;
  const [authorityPublicKey, signers] = isSigner(freezeAuthority) ? [freezeAuthority.publicKey, [freezeAuthority]] : [freezeAuthority, multiSigners];
  const tokenAddressOrAta = tokenAddress !== null && tokenAddress !== void 0 ? tokenAddress : findAssociatedTokenAccountPda(mintAddress, tokenOwner);
  return TransactionBuilder.make().add({
    instruction: createThawAccountInstruction(tokenAddressOrAta, mintAddress, authorityPublicKey, multiSigners, tokenProgram),
    signers,
    key: (_params$instructionKe = params.instructionKey) !== null && _params$instructionKe !== void 0 ? _params$instructionKe : 'thawTokens'
  });
};

export { _thawTokensBuildersClient, _thawTokensClient, thawTokensBuilder, thawTokensOperation, thawTokensOperationHandler };
