// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/accounts.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var mplAuctionHouse = require('@metaplex-foundation/mpl-auction-house');
var Account = require('../../types/Account.cjs');

const parseAuctioneerAccount = Account.getAccountParsingFunction(mplAuctionHouse.Auctioneer);
const toAuctioneerAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.Auctioneer);
const parseAuctionHouseAccount = Account.getAccountParsingFunction(mplAuctionHouse.AuctionHouse);
const toAuctionHouseAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.AuctionHouse);
const parseListingReceiptAccount = Account.getAccountParsingFunction(mplAuctionHouse.ListingReceipt);
const toListingReceiptAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.ListingReceipt);
const parseBidReceiptAccount = Account.getAccountParsingFunction(mplAuctionHouse.BidReceipt);
const toBidReceiptAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.BidReceipt);
const parsePurchaseReceiptAccount = Account.getAccountParsingFunction(mplAuctionHouse.PurchaseReceipt);
const toPurchaseReceiptAccount = Account.getAccountParsingAndAssertingFunction(mplAuctionHouse.PurchaseReceipt);

exports.parseAuctionHouseAccount = parseAuctionHouseAccount;
exports.parseAuctioneerAccount = parseAuctioneerAccount;
exports.parseBidReceiptAccount = parseBidReceiptAccount;
exports.parseListingReceiptAccount = parseListingReceiptAccount;
exports.parsePurchaseReceiptAccount = parsePurchaseReceiptAccount;
exports.toAuctionHouseAccount = toAuctionHouseAccount;
exports.toAuctioneerAccount = toAuctioneerAccount;
exports.toBidReceiptAccount = toBidReceiptAccount;
exports.toListingReceiptAccount = toListingReceiptAccount;
exports.toPurchaseReceiptAccount = toPurchaseReceiptAccount;
