// @metaplex-foundation/js/dist/cjs/plugins/auctionHouseModule/loadListing.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Nft = require('../nftModule/Nft.cjs');
var Operation = require('../../types/Operation.cjs');
var Amount = require('../../types/Amount.cjs');

// Operation
// -----------------

const Key = 'LoadListingOperation';
const loadListingOperation = Operation.useOperation(Key);
// -----------------
// Handler
// -----------------
const loadListingOperationHandler = {
  handle: async (operation, metaplex, scope) => {
    const {
      lazyListing,
      loadJsonMetadata = true,
      commitment
    } = operation.input;
    const asset = await metaplex.nfts().findByMetadata(lazyListing.metadataAddress, {
      tokenOwner: lazyListing.sellerAddress,
      commitment,
      loadJsonMetadata
    }).run(scope);
    Nft.assertNftOrSftWithToken(asset);
    return { ...lazyListing,
      model: 'listing',
      lazy: false,
      asset,
      tokens: Amount.amount(lazyListing.tokens, asset.mint.currency)
    };
  }
};

exports.loadListingOperation = loadListingOperation;
exports.loadListingOperationHandler = loadListingOperationHandler;
